### PR TITLE
feat(remote-build): new command logic

### LINKED ISSUE #4324: remote-build: new command logic
### What needs to get done

- [ ] Implement the new work area, where the WorkTree is not reused and removed after the build completes.
    - Each launchpad build has a unique build-id and WorkTree.  In the legacy code, WorkTrees are stored by build-id in `~/.local/share/snapcraft/remote-build/<build-id>/`.
    - In the new code, WorkTrees will be stored in the work area  `~/.cache/snapcraft/remote-build/<build-id>/`.  When the build completes successfully, the WorkTree will be removed.  WorkTrees are not reused.
- [ ] use new `_generate_build_id()` from #4322 
- [ ] linting and typing updates (e.g. use f-strings)
- [ ] move snapcraft business logic to `snapcraft/commands/remote.py`

### Why it needs to get done

This allows a user to inspect the contents of the WorkTree if the build fails and minimizes disk usage.

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

- Migrate the command logic used to orchestrate a remote build.
  - Some of this code was moved to a new class, `RemoteBuild`
- Store a copy of the project directory in the local cache
- Clean up the cache after a build completes
- Typing and linting updates

Builds on #4390, so "only" the last 8 commits are for this PR (update: #4390 was landed and this PR is rebased onto main).  The new remote-build should be functional, so I encourage you to give it try.

Fixes #4324 
(CRAFT-1979)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -65,7 +67,13 @@ plugins = [
 
 [tool.pyright]
 include = ["snapcraft", "tests"]
-exclude = ["tests/legacy", "tests/spread", "build"]
+exclude = [
+    "build",
+    # launchpadlib is not typed
+    "snapcraft/remote/launchpad.py",
```
[the remark this answers is not in the document]
Comment (lengau) [reply pyproject.toml:73]: Personal opinion: I'd rather have a lot of type checker noise in `launchpad.py` than exclude the file entirely

File: snapcraft/remote/utils.py [hunk 2]
```diff
@@ -76,3 +98,57 @@ def _compute_hash(directory: Path) -> str:
 
     all_hashes = "".join(hashes).encode()
     return md5(all_hashes).hexdigest()  # noqa: S324 (insecure-hash-function)
+
+
+def humanize_list(
```
Comment (lengau) [inline snapcraft/remote/utils.py:103]: This replicates `snapcraft.utils.humanize_list`.

Bigger picture: This is probably a good utility function to put into `craft-application`

**Review by lengau — CHANGES_REQUESTED** [review]

File: snapcraft/remote/utils.py (hunk 2 above)
Comment (lengau) [reply snapcraft/remote/utils.py:103]: Based on that, your call.

If remote build is going to become its own library that doesn't depend on `craft-application` (and given how we have this function in so many places), we may even want to make a little `craft-utils` library for all the little miscellany we have. (`humanize_list`, possibly architecture related stuff, maybe even our base error class. But I would want to keep it without non-stdlib dependencies if possible.)

Comment (sergiusens) [reply snapcraft/remote/utils.py:103]: craft cli is all about presentation though, utils risks being a dumpster of things we do not know or do not want to think of where to put

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mr-cal) [reply pyproject.toml]: Well, I feel silly for excluding this file in the first place.  A few typing fixes and `casts` (thanks @lengau for the hint) solved dozens of linter errors.

To answer your question Claudio, `#type: ignore` in the import line doesn't work.

File: snapcraft/commands/remote.py [hunk 3]
```diff
@@ -279,3 +308,53 @@ def _get_effective_base(self) -> str:
...
+        :returns: A list of architectures.
+
+        :raises SnapcraftError: If `--build-on` was provided and architectures are
+        defined in the project's snapcraft.yaml.
+        """
+        project_architectures = self._get_project_architectures()
+        if project_architectures and self._parsed_args.build_for:
+            raise SnapcraftError(
+                "Cannot use `--build-on` because architectures are already defined in "
+                "the snapcraft.yaml."
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/commands/remote.py:349]: I think that the phrases `in snapcraft.yaml` or `in the snapcraft.yaml file` are both fine (I mixed the two here).  I'll use your suggestion.

File: snapcraft/commands/remote.py [hunk 4]
```diff
@@ -279,3 +308,53 @@ def _get_effective_base(self) -> str:
...
+    def _determine_architectures(self) -> List[str]:
+        """Determine architectures to build for.
+
+        The build architectures can be set via the `--build-on` parameter or determined
+        from the build-on architectures listed in the project's snapcraft.yaml.
+
+        :returns: A list of architectures.
+
+        :raises SnapcraftError: If `--build-on` was provided and architectures are
+        defined in the project's snapcraft.yaml.
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/commands/remote.py:343]: Like this?
```
architectures:
  - build-on: [amd64]
    build-for: [amd64]
  - build-on: [amd64]
    build-for: [arm64]
```

Unfortunately, we can't define architectures in a `snapcraft.yaml` file and request for Launchpad to only build some variants.

File: snapcraft/remote/utils.py (hunk 2 above)
Comment (mr-cal) [reply snapcraft/remote/utils.py:103]: I decided to leave it as-is.

Comment (lengau) [reply snapcraft/remote/utils.py:103]: Yeah, that is a concern. `craft-cli` is also a good spot, but I wouldn't want to make `craft-parts` depend on it (and it currently uses its own `humanize_list` in 9 places). Although if we're reducing having it in all the apps + craft-parts to just craft-parts and craft-cli that's already a huge improvement.

I've made this to start out with: https://github.com/canonical/craft-cli/issues/183

File: snapcraft/remote/launchpad.py [hunk 5]
```diff
@@ -102,7 +102,7 @@ def __init__(
         self._project_name = project_name
 
         self._lp: Launchpad = self._login()
-        self.user = self._lp.me.name
+        self.user = self._lp.me.name  # type: ignore
```
Comment (lengau) [inline snapcraft/remote/launchpad.py:105]:
```suggestion
        self.user = cast(str, self._lp.me.name)
```

should work too

File: snapcraft/remote/launchpad.py [hunk 6]
```diff
@@ -160,28 +160,30 @@ def _fetch_artifacts(self, snap: Entry) -> None:
             self._download_build_artifacts(build)
             self._download_log(build)
 
-    def _get_builds_collection_entry(self, snap: Entry) -> Optional[Entry]:
+    def _get_builds_collection_entry(self, snap: Optional[Entry]) -> Optional[Entry]:
         logger.debug("Fetching builds collection information from Launchpad...")
-        url = snap.builds_collection_link
-        return self._lp_load_url(url)
+        url = snap.builds_collection_link  # type: ignore
+        return self._lp_load_url(cast(str, url))
```
Comment (lengau) [inline snapcraft/remote/launchpad.py:166]:
```suggestion
        url = cast(str, snap.builds_collection_link)
        return self._lp_load_url(url)
```

**Review by lengau — APPROVED** [review]
I've made two suggestions about typing in launchpad.py, but I haven't actually tried them to know whether both mypy *and* pyright will behave with them, and frankly there's no need for them.

File: snapcraft/remote/launchpad.py (hunk 5 above)
Comment (mr-cal) [reply snapcraft/remote/launchpad.py:105]: I tried that, but the problem is with `_lp` being incorrectly typed:
```
error: Cannot access member "name" for type "NamedOperation"
   Member "name" is unknown (reportGeneralTypeIssues)
error: "name" is not a known member of "None"  (reportOptionalMemberAccess)
```

I can rewrite it as:
```
me = self._lp.me
self.user = me.name
```
and the same two warnings persist for the first line.

So I left it as-is.

File: snapcraft/remote/launchpad.py (hunk 6 above)
Comment (mr-cal) [reply snapcraft/remote/launchpad.py]: Oo thanks.  You got me on the right track, check it out now.

File: snapcraft/remote/launchpad.py (hunk 5 above)
Comment (lengau) [reply snapcraft/remote/launchpad.py:105]: Well that's disappointing... Thanks for checking!

**Review by lengau — APPROVED** [review]

File: snapcraft/remote/launchpad.py (hunk 6 above)
Comment (lengau) [reply snapcraft/remote/launchpad.py]: Fantastic!

**Review by sergiusens — APPROVED** [review]

**Review by cmatsuoka — APPROVED** [review]
Awesome, thanks!

**Pull request merged**